packages/libs/deer/src/lib.rs

```diff
@@ -6,4 +6,212 @@
+    ///
+    /// This method also enables a default trait implementation for all primitives types.
+    ///
+    /// > **Hint**: Do not try to coerce values like `"1"`, `"0"`, `true`, `false` to numbers, this
+    /// > is highly discouraged. The [`Deserialize`] implementation of a future version of `deer`
+    /// > would implement that functionality.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type number
+    fn number(self) -> Result<Number, Self::Error>;
+
+    /// Deserialize a [`String`] value.
+    ///
+    /// `deer` (in contrast to [`serde`]) does not support `&str`, this is because `deer` only
+    /// handles owned data, this comes with some overhead, but enables easier implementation.
+    ///
+    /// > **Hint**: Do not try to cerce values to string, this is highly discouraged and only
+    /// > [`Deserialize`] or a future version of `deer` should implement that functionality, if
+    /// > desired.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type string
+    ///
+    /// [`serde`]: https://serde.rs/
+    fn string(self) -> Result<String, Self::Error>;
+
+    /// Deserialize a `Array`
+    ///
+    /// This should not directly deserialize into a `Vec<T>`, but return a type that implements
+    /// `ArrayAccess`, this type then iterates over all values.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type array
+    fn array(self) -> Result<Self::Array, Self::Error>;
+
+    /// Deserialize a `Object`
+    ///
+    /// This should not directly deserialize into a object, but instead return a type that
+    /// implements `ObjectAccess`, this type will then go through all entries.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type object
+    fn object(self) -> Result<Self::Object, Self::Error>;
+
+    derive_from_number!([
+        i8; to_i8,
+        i16; to_i16,
+        i32; to_i32,
+        i64; to_i64,
+        isize; to_isize,
+        u8; to_u8,
+        u16; to_u16,
+        u32; to_u32
+    ]);
+
+    derive_from_number!([
+        i128; to_i128,
+        u64; to_u64,
+        u128; to_u128,
+        usize; to_usize
+    ]);
+}
+
+/// A **data-structure** that can be deserialized from any format supported by deer.
+///
+/// `deer` provides [`Deserialize`] implementations for many Rust primitives and standard library
+/// types.
+///
+/// Additionally `deer` provides a derive macro which can automatically generate the trait.
+///
+/// In rare cases it may be necessary to implement [`Deserialize`] manually, in that case you can
+/// use the automatically generated output
+/// (which can be displayed with tools like [cargo-expand](https://github.com/dtolnay/cargo-expand))
+/// as a template. The macro generates human readable code which can be used as template.
+pub trait Deserialize: Sized {
+    /// The name of the type, this is used during error reporting to display a correct name.
+    fn name() -> &'static str;
```

Comment:
What would be returned if the type is anonymous? (e.g. the elements of an array)

Reply:
It would always be the name of the struct, so for `Vec<T>`, it would be `Vec`. For `Example<T, U>`, it would be `Example`.

Reply:
With the visitor pattern now in place, would it make sense to move it to `Visitor` instead? (`serde` uses the `expected` function)

Reply:
Yes (I actually just finished the PR for exactly that^^)

Reply:
fixed in [cc6711d](https://github.com/hashintel/hash/pull/1170/commits/cc6711d6d412b3e0037fe33c17e5f0d126f70262)